invenio_drafts_resources/records/api.py

```diff
@@ -93,12 +124,90 @@ def register(self):
 class Draft(Record):
     """Draft base API for metadata creation and manipulation."""
 
-    # WHY: We want to force the model_cls to be specified by the user
-    # No default one is given, only the base.
+    #
+    # Configuration to be set by a subclass
+    #
+
+    #: The record's SQLAlchemy model class. Must be set by the subclass.
     model_cls = None
+    #: The parent state's SQLAlchemy model class. Must be set by the subclass.
+    versions_model_cls = None
+    #: The parent record's API class. Must be set by the subclass.
+    parent_record_cls = None
 
+    #
+    # System fields
+    #
+
+    #: The internal persistent identifier. Records and drafts share UUID.
     pid = PIDField('id', provider=DraftRecordIdProviderV2, delete=False)
 
+    #: The parent record - the draft is responsible for creating the parent.
+    parent = ParentField(
+        ParentRecord, create=True, soft_delete=False, hard_delete=True)
+
+    #: Version relationship
+    versions = VersionsField(create=True, set_next=True)
+
+    #: The expiry date of the draft.
     expires_at = ModelField()
 
+    #: Revision id of record from which this draft was created.
     fork_version_id = ModelField()
+
+    @classmethod
+    def new_version(cls, record):
+        """Create a draft for a new version of a record.
+
+        The caller is responsible for:
+        1) checking if a draft for a new version already exists
+        2) moving the record data into the draft data.
+        """
+        return cls.create(
+            {},
+            # We create a new id, because this is for a new version.
+            id=uuid.uuid4(),
+            # Links the draft with the same parent (i.e. a new version).
+            parent=record.parent,
+            versions=record.versions,
+            # New drafts without a record (i.e. unpublished drafts) must set
+            # the fork version id to None.
+            fork_version_id=None,
+        )
+
+    @classmethod
+    def edit(cls, record):
+        """Create a draft for editing an existing version of a record."""
+        try:
+            # We soft-delete a draft once it has been published, in order to
+            # keep the version_id counter around for optimistic concurrency
+            # control (both for ES indexing and for REST API clients)
+            draft = cls.get_record(record.id, with_deleted=True)
+            if draft.is_deleted:
+                draft.undelete()
```

Comment:
We issue a soft-delete (i.e. draft stays in the table), and then issue a delete to Elasticsearch. Elasticsearch then keeps around the UUID + version id of the deleted record for 1 min. That means, that if in that 1 min, somebody goes and press the edit button again (i.e. we're creating a draft), we need to index the same UUID again. If the version id we send is lower than the one we used for the ES delete operation (e.g. in case we hard-deleted the record), ES will refuse the operation because it thinks it's an old operation that arrived late. That's why we need to keep the draft version id around, so that we send an ES index operation with a version id that's higher than the version id we used in the ES delete.......the way we keep the version_id around is by simply soft-deleting the draft, but we coudl also just have stored the version id somewhere else.

Reply:
Ah, got it. So ES has its own "versioning" mechanism that it uses to determine that insertion of a deleted document with a more recent "version id" is fine. (It's not that we are telling ES to cancel deletion. We just tell it to insert one with higher version and that is fine whether the 1min mark has elapsed or not).